File: src/long_pull.py
"""Copyright 2012, System Insights, Inc.

   Licensed under the Apache License, Version 2.0 (the "License");
   you may not use this file except in compliance with the License.
   You may obtain a copy of the License at

       http://www.apache.org/licenses/LICENSE-2.0

   Unless required by applicable law or agreed to in writing, software
   distributed under the License is distributed on an "AS IS" BASIS,
   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
   See the License for the specific language governing permissions and
   limitations under the License."""
    
import socket, select, re
import scanner
import httplib

class LongPullException(Exception):
    pass

class LongPull:
    def __init__(self, response):
        self._response = response
        self._buffer = ''

    def _read_all(self, size):
        globs = [self._buffer]
        amt = size - len(self._buffer)


        while amt > 0:
            ready, d1, d2 = select.select([self._socket], [], [])
            if len(ready) > 0:
                glob = self._socket.recv(amt)
                if not glob:
                    raise LongPullException(''.join(globs), amt)
                amt -= len(glob)
                globs.append(glob)

        return ''.join(globs)


    def _read_chunk(self):
        # This should always occur on a chunk boundry, so we should never
        # need to store a lot of text in the buffer.

        text = ''
        chunk_size = None

        if len(self._buffer) < 32:
            ready, d1, d2 = select.select([self._socket], [], [])
            if len(ready) == 0:
                raise httplib.NotConnected()

            text = self._socket.recv(32)
            if len(text) == 0:
                self._socket.close()
                raise httplib.NotConnected()

        self._buffer += text
        eol = self._buffer.find('\r\n')
        if eol >= 0:
            line = self._buffer[:eol]
            i = line.find(';')
            if i >= 0:
                line = line[:i] # strip chunk-extensions
            try:
                chunk_size = int(line, 16) + 2
            except ValueError:
                # close the connection as protocol synchronisation is
                # probably lost
                self._socket.close()
                raise httplib.IncompleteRead(line)
            self._buffer = self._buffer[(eol + 2):]

        chunk = self._read_all(chunk_size)
        self._buffer = chunk[chunk_size:]

        return chunk[:(chunk_size - 2)]


    def long_pull(self, callback, user_data = None):
        fileno = self._response.fileno()
        self._socket = socket.fromfd(fileno, socket.AF_INET, socket.SOCK_STREAM)
        self._socket.setblocking(False)
        content_type = dict(self._response.getheaders())['content-type']
        match = re.search('boundary=([0-9A-Fa-f]+)', content_type)
        if not match:
            raise LongPullException('Cannot find boundary in content-type')

        boundary = '--' + match.group(1)
        boundary_pat = re.compile('^' + boundary)
        header = True
        length = len(boundary)
        document = scanner.Scanner('')
        while True:
            chunk = self._read_chunk()
            document.string += chunk

            while document.rest_len() >= length:
                if header:
                    if not document.check(boundary_pat):
                        print "Framing error!"
                        raise LongPullException('Framing error')

                    head = document.scan_until('\r\n\r\n')
                    mime_headers = head.split('\r\n')
                    values = dict([(v.lower().split(':')) for v in mime_headers[1:] if v.find(':') > 0])
                    header = False
                    try:
                        length = int(values['content-length'])
                    except ValueError:
                        raise LongPullException('Cannot get length from mime header: ' + mime_headers)

                else:
                    rest = document.rest()
                    body = rest[:length]

                    document.reset()
                    document.string = rest[length:]

                    callback(body)

                    length = len(boundary)
                    header = True

if __name__ == "__main__":
    conn = httplib.HTTPConnection('agent.mtconnect.org')
    conn.request("GET", "/sample?interval=1000&count=1000")
    response = conn.getresponse()

    lp = LongPull(response)
    def callback(chunk):
        print chunk

    lp.long_pull(callback)

File: mac_robot_sim/Robot/long_pull.rb
# Copyright 2012, System Insights, Inc.
#
#   Licensed under the Apache License, Version 2.0 (the "License");
#   you may not use this file except in compliance with the License.
#   You may obtain a copy of the License at
#
#       http://www.apache.org/licenses/LICENSE-2.0
#
#    Unless required by applicable law or agreed to in writing, software
#    distributed under the License is distributed on an "AS IS" BASIS,
#    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#    See the License for the specific language governing permissions and
#    limitations under the License.


require 'net/http'
require 'strscan'

class LongPull
  class Timeout < RuntimeError; end
  
  def initialize(http)
    @http = http
  end

  def long_pull(uri, timeout = 2.0, &block)
    @http.request_get(uri) do |res|
      content_type, = res.get_fields('content-type')
      if content_type !~ /boundary=([^;]+)/
        raise "No boundry: #{res.body}"
      end
      
      boundary = "--#{$1}"
      document = StringScanner.new(String.new)
      header = true
      length = boundary.length
      
      # Create a timeout thread to kill the read in case the agent becomes unresponsive...
      ct = Thread.current
      chunk_time = Time.now
      tot = Thread.start {
        while true
          nap = timeout - (Time.now - chunk_time)
          if nap > 0.0
            sleep nap 
          else
            ct.raise Timeout, 'Read timed out'
          end
        end
      }
      
      begin
        res.read_body do |chunk|
          next if chunk.empty?
          document << chunk

          while document.rest_size >= length
            if header
              if !document.check(/^#{boundary}/)
                puts document.rest
                raise "Framing error"
              end
                        
              break unless head = document.scan_until(/\r\n\r\n/)
              bound, *rest = head.split(/\r\n/)
            
              fields = Hash[*rest.map { |s| s.split(/:\s*/) }.flatten]
              length = fields['Content-length'].to_i
              header = false
            else
              # We have now received a complete chunk from the server, so we'll mark time here to 
              # make the timeout thread happy.
              chunk_time = Time.now
              rest = document.rest
            
              # Slice off the chunk we need
              body = rest[0, length]
            
              document.clear
              document = StringScanner.new(rest[length..-1])
              
              block.call(body)
              # puts body
            
              # New message
              length = boundary.length
              header = true
            end
          end
        end
      ensure
        tot.kill
        # Make sure it exited, don't want it killing anything else...
        tot.join
      end
    end
  end
end
